WFS20 ComparisonOperatorTests Request Contains Invalid Parameter

To OGC Team,

The following test failed:

property Is Greater Than_match Any

Class Name:ComparisonOperatorTests
Method Name:propertyIsGreaterThan_matchAny29
Path:org/opengis/cite/iso19142/basic/filter/ComparisonOperatorTests

Error:
java.lang.AssertionError: Unexpected HTTP status code. expected <br>[200] but found <br>[400]

This occurred from an invalid REST Request that was generated by the Validator.

The following request was generated: 

http://agsinspire1.esri.com/arcgis/rest/services/PSTest/MapServer/exts/InspireFeatureDownload/service?service=WFS&version=2.0.0&request=GetFeature&count=10&typenames=ns43:ProtectedSite&namespaces=xmlns(xml,http://www.w3.org/XML/1998/namespace),xmlns(ns43,http://inspire.ec.europa.eu/schemas/ps/4.0),xmlns(wfs,http://www.opengis.net/wfs/2.0)&filter=%3CFilter%20xmlns%3D%22http%3A%2F%2Fwww.opengis.net%2Ffes%2F2.0%22%20xmlns%3Ans43%3D%22http%3A%2F%2Finspire.ec.europa.eu%2Fschemas%2Fps%2F4.0%22%20xmlns%3Awfs%3D%22http%3A%2F%2Fwww.opengis.net%2Fwfs%2F2.0%22%3E%3CPropertyIsGreaterThan%20matchAction%3D%22Any%22%20matchCase%3D%22true%22%3E%3CLiteral%3E1799-12-31T00%3A03%3A58-04%3A56%3A02%3C%2FLiteral%3E%3CValueReference%20xmlns%3Atns%3D%22http%3A%2F%2Finspire.ec.europa.eu%2Fschemas%2Fps%2F4.0%22%3Etns%3AlegalFoundationDate%3C%2FValueReference%3E%3C%2FPropertyIsGreaterThan%3E%3C%2FFilter%3E

Decoded request:

http://agsinspire1.esri.com/arcgis/rest/services/PSTest/MapServer/exts/InspireFeatureDownload/service?service=WFS&version=2.0.0&request=GetFeature&count=10&typenames=ns43:ProtectedSite&namespaces=xmlns(xml,http://www.w3.org/XML/1998/namespace),xmlns(ns43,http://inspire.ec.europa.eu/schemas/ps/4.0),xmlns(wfs,http://www.opengis.net/wfs/2.0)&filter=<Filter xmlns="http://www.opengis.net/fes/2.0" xmlns:ns43="http://inspire.ec.europa.eu/schemas/ps/4.0" xmlns:wfs="http://www.opengis.net/wfs/2.0"><PropertyIsGreaterThan matchAction="Any" matchCase="true"><Literal>1799-12-31T00:03:58-04:56:02</Literal><ValueReference xmlns:tns="http://inspire.ec.europa.eu/schemas/ps/4.0">tns:legalFoundationDate</ValueReference></PropertyIsGreaterThan></Filter>

The <Literal> datetime value seems to be the issue.
1799-12-31T00:03:58-04:56:02

It seems the datetime format is not correct.

If I change the datetime to the following the request succeeds.
1799-12-31T00:03:58


Can someone verify this?

[PS_GetCapabilities.zip](https://github.com/opengeospatial/teamengine/files/5557047/PS_GetCapabilities.zip)

--- a/src/main/java/org/opengis/cite/iso19142/basic/filter/ComparisonOperatorTests.java
+++ b/src/main/java/org/opengis/cite/iso19142/basic/filter/ComparisonOperatorTests.java
@@ -554,6 +554,9 @@ public class ComparisonOperatorTests extends QueryFilterFixture {
             if (GregorianCalendar.class.isInstance(objValues[i])) {
                 GregorianCalendar cal = (GregorianCalendar) objValues[i];
                 values[i] = tmFormatter.format(cal.toZonedDateTime());
+                if (values[i].contains("+") && values[i].length() > 25) {
+                    values[i] = values[i].substring(0, 25);
+                }
             } else {
                 values[i] = DatatypeConverter.printDecimal(new BigDecimal(objValues[i].toString()));
             }

--- a/src/main/java/org/opengis/cite/iso19142/basic/filter/ComparisonOperatorTests.java
+++ b/src/main/java/org/opengis/cite/iso19142/basic/filter/ComparisonOperatorTests.java
@@ -548,15 +548,12 @@ public class ComparisonOperatorTests extends QueryFilterFixture {
             }
             Arrays.sort(objValues);
         }
-        DateTimeFormatter tmFormatter = (values[0].indexOf('T') > 0) ? DateTimeFormatter.ISO_OFFSET_DATE_TIME
+        DateTimeFormatter tmFormatter = (values[0].indexOf('T') > 0) ? DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssXXX")
                 : DateTimeFormatter.ISO_DATE;
         for (int i = 0; i < values.length; i++) {
             if (GregorianCalendar.class.isInstance(objValues[i])) {
                 GregorianCalendar cal = (GregorianCalendar) objValues[i];
                 values[i] = tmFormatter.format(cal.toZonedDateTime());
-                if (values[i].contains("+") && values[i].length() > 25) {
-                    values[i] = values[i].substring(0, 25);
-                }
             } else {
                 values[i] = DatatypeConverter.printDecimal(new BigDecimal(objValues[i].toString()));
             }

--- a/src/main/java/org/opengis/cite/iso19142/basic/filter/ComparisonOperatorTests.java
+++ b/src/main/java/org/opengis/cite/iso19142/basic/filter/ComparisonOperatorTests.java
@@ -548,6 +548,8 @@ public class ComparisonOperatorTests extends QueryFilterFixture {
             }
             Arrays.sort(objValues);
         }
+        // Earlier we using ISO_OFFSET_DATE_TIME but offset is appended with seconds due to that test is getting failed.
+        // So we have added a custom datetime formatter pattern(yyyy-MM-dd'T'HH:mm:ssXXX) which will ignore seconds from offset.
         DateTimeFormatter tmFormatter = (values[0].indexOf('T') > 0) ? DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssXXX")
                 : DateTimeFormatter.ISO_DATE;
         for (int i = 0; i < values.length; i++) {

